datadog/resource_datadog_role_test.go

```diff
@@ -0,0 +1,187 @@
+package datadog
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+)
+
+func TestAccDatadogRole_CreateUpdate(t *testing.T) {
+	accProviders, clock, cleanup := testAccProviders(t, initRecorder(t))
+	rolename := strings.ToLower(uniqueEntityName(clock, t))
+	defer cleanup(t)
+	accProvider := testAccProvider(t, accProviders)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    accProviders,
+		CheckDestroy: testAccCheckDatadogRoleDestroy(accProvider),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckDatadogRoleConfig(rolename),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckDatadogRoleExists(accProvider, "datadog_role.foo"),
+					resource.TestCheckResourceAttr("datadog_role.foo", "name", rolename),
+					resource.TestCheckResourceAttr("datadog_role.foo", "permission.#", "2"),
+					testCheckRolePermission(
+						"datadog_role.foo",
+						"data.datadog_permissions.foo",
+						"permissions.admin",
+					),
+					testCheckRolePermission(
+						"datadog_role.foo",
+						"data.datadog_permissions.foo",
+						"permissions.standard",
+					),
+				),
+			},
+			{
+				Config: testAccCheckDatadogRoleConfigUpdated(rolename + "updated"),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckDatadogRoleExists(accProvider, "datadog_role.foo"),
+					resource.TestCheckResourceAttr("datadog_role.foo", "permission.#", "2"),
+					testCheckRolePermission(
+						"datadog_role.foo",
+						"data.datadog_permissions.foo",
+						"permissions.standard",
+					),
+					testCheckRolePermission(
+						"datadog_role.foo",
+						"data.datadog_permissions.foo",
+						"permissions.logs_read_index_data",
+					),
+				),
+			},
+			{
+				Config: testAccCheckDatadogRoleConfigNoPerm(rolename + "noperm"),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckDatadogRoleExists(accProvider, "datadog_role.foo"),
+					resource.TestCheckResourceAttr("datadog_role.foo", "permission.#", "0"),
+				),
+			},
+		},
+	})
+}
+func TestAccDatadogRole_RestrictedPerm(t *testing.T) {
```

Comment:
Super small nit, but s/restricted/invalid since this permission doesn't exist
```suggestion
func TestAccDatadogRole_InvalidPerm(t *testing.T) {
```